File: PersonalFinanceTraker/PersonalFinanceTraker/Utilities/XLSXExportService.swift
//
//  XLSXExportService.swift
//  PersonalFinanceTraker
//

import Foundation
import ZIPFoundation

enum XLSXWriteError: LocalizedError {
    case archiveFailed

    var errorDescription: String? {
        "Couldn't create the Excel file."
    }
}

/// Minimal .xlsx writer for exporting transactions. Same columns as
/// CSVExportService so the two export formats stay in sync.
///
/// ponytail: hand-rolled OOXML (5 XML parts in a ZIP) instead of an Excel
/// library — CoreXLSX is read-only. Dates are exported as plain text and
/// amounts as bare numbers; add styles.xml numFmts and date serials if
/// anyone needs Excel-native date/currency cells.
enum XLSXExportService {

    static func generateXLSX(from transactions: [TransactionSnapshot]) throws -> Data {
        let parts: [(path: String, xml: String)] = [
            ("[Content_Types].xml", contentTypesXML),
            ("_rels/.rels", rootRelsXML),
            ("xl/workbook.xml", workbookXML),
            ("xl/_rels/workbook.xml.rels", workbookRelsXML),
            ("xl/styles.xml", stylesXML),
            ("xl/worksheets/sheet1.xml", sheetXML(from: transactions)),
        ]

        let archive = try Archive(data: Data(), accessMode: .create)
        for part in parts {
            let data = Data(part.xml.utf8)
            try archive.addEntry(
                with: part.path,
                type: .file,
                uncompressedSize: Int64(data.count)
            ) { position, size in
                data.subdata(in: Int(position)..<(Int(position) + size))
            }
        }
        guard let data = archive.data else { throw XLSXWriteError.archiveFailed }
        return data
    }

    // MARK: - Sheet

    private static let headers = ["Date", "Amount", "Currency", "Category", "Note", "Type"]

    private static func sheetXML(from transactions: [TransactionSnapshot]) -> String {
        let dateFormatter = DateFormatter()
        dateFormatter.dateFormat = "yyyy-MM-dd HH:mm:ss"

        var rows: [String] = [row(number: 1, cells: headers.map { .text($0) })]
        for (idx, tx) in transactions.enumerated() {
            rows.append(row(number: idx + 2, cells: [
                .text(dateFormatter.string(from: tx.timestamp)),
                .number("\(tx.amount)"),
                .text(tx.currencyCode),
                .text(tx.category),
                .text(tx.note),
                .text(tx.amount >= 0 ? "Income" : "Expense"),
            ]))
        }

        return """
        <?xml version="1.0" encoding="UTF-8" standalone="yes"?>
        <worksheet xmlns="http://schemas.openxmlformats.org/spreadsheetml/2006/main"><sheetData>\(rows.joined())</sheetData></worksheet>
        """
    }

    private enum CellValue {
        case text(String)
        case number(String)
    }

    /// Emits explicit cell references (A1, B2, …) — readers, including our own
    /// XLSXWorkbook importer, rely on them for column placement.
    private static func row(number: Int, cells: [CellValue]) -> String {
        let columns = ["A", "B", "C", "D", "E", "F"]
        let xml = cells.enumerated().map { idx, cell in
            let ref = "\(columns[idx])\(number)"
            switch cell {
            case .text(let value):
                return #"<c r="\#(ref)" t="inlineStr"><is><t xml:space="preserve">\#(escapeXML(value))</t></is></c>"#
            case .number(let value):
                return #"<c r="\#(ref)"><v>\#(value)</v></c>"#
            }
        }.joined()
        return #"<row r="\#(number)">\#(xml)</row>"#
    }

    private static func escapeXML(_ text: String) -> String {
        text
            .replacingOccurrences(of: "&", with: "&amp;")
            .replacingOccurrences(of: "<", with: "&lt;")
            .replacingOccurrences(of: ">", with: "&gt;")
    }

    // MARK: - Boilerplate package parts

    private static let contentTypesXML = """
    <?xml version="1.0" encoding="UTF-8" standalone="yes"?>
    <Types xmlns="http://schemas.openxmlformats.org/package/2006/content-types"><Default Extension="rels" ContentType="application/vnd.openxmlformats-package.relationships+xml"/><Default Extension="xml" ContentType="application/xml"/><Override PartName="/xl/workbook.xml" ContentType="application/vnd.openxmlformats-officedocument.spreadsheetml.sheet.main+xml"/><Override PartName="/xl/worksheets/sheet1.xml" ContentType="application/vnd.openxmlformats-officedocument.spreadsheetml.worksheet+xml"/><Override PartName="/xl/styles.xml" ContentType="application/vnd.openxmlformats-officedocument.spreadsheetml.styles+xml"/></Types>
    """

    private static let rootRelsXML = """
    <?xml version="1.0" encoding="UTF-8" standalone="yes"?>
    <Relationships xmlns="http://schemas.openxmlformats.org/package/2006/relationships"><Relationship Id="rId1" Type="http://schemas.openxmlformats.org/officeDocument/2006/relationships/officeDocument" Target="xl/workbook.xml"/></Relationships>
    """

    private static let workbookXML = """
    <?xml version="1.0" encoding="UTF-8" standalone="yes"?>
    <workbook xmlns="http://schemas.openxmlformats.org/spreadsheetml/2006/main" xmlns:r="http://schemas.openxmlformats.org/officeDocument/2006/relationships"><sheets><sheet name="Transactions" sheetId="1" r:id="rId1"/></sheets></workbook>
    """

    private static let workbookRelsXML = """
    <?xml version="1.0" encoding="UTF-8" standalone="yes"?>
    <Relationships xmlns="http://schemas.openxmlformats.org/package/2006/relationships"><Relationship Id="rId1" Type="http://schemas.openxmlformats.org/officeDocument/2006/relationships/worksheet" Target="worksheets/sheet1.xml"/><Relationship Id="rId2" Type="http://schemas.openxmlformats.org/officeDocument/2006/relationships/styles" Target="styles.xml"/></Relationships>
    """

    private static let stylesXML = """
    <?xml version="1.0" encoding="UTF-8" standalone="yes"?>
    <styleSheet xmlns="http://schemas.openxmlformats.org/spreadsheetml/2006/main"><fonts count="1"><font><sz val="12"/><name val="Calibri"/></font></fonts><fills count="1"><fill><patternFill patternType="none"/></fill></fills><borders count="1"><border/></borders><cellStyleXfs count="1"><xf numFmtId="0" fontId="0" fillId="0" borderId="0"/></cellStyleXfs><cellXfs count="1"><xf numFmtId="0" fontId="0" fillId="0" borderId="0" xfId="0"/></cellXfs></styleSheet>
    """
}

File: Android/app/src/main/java/com/rizzog99/personalfinancetracker/domain/export/XlsxExportService.kt
package com.rizzog99.personalfinancetracker.domain.export

import com.rizzog99.personalfinancetracker.domain.transaction.FinanceTransaction
import java.io.ByteArrayOutputStream
import java.util.zip.ZipEntry
import java.util.zip.ZipOutputStream
import java.time.ZoneId
import java.time.format.DateTimeFormatter
import java.util.Locale

/** Minimal standards-compliant XLSX writer using inline strings; keeps Android exports readable by Excel and Numbers. */
object XlsxExportService {
    private val dateFormatter = DateTimeFormatter.ofPattern("yyyy-MM-dd HH:mm:ss", Locale.ROOT).withZone(ZoneId.systemDefault())
    fun generate(transactions: List<FinanceTransaction>): ByteArray {
        val rows = listOf(listOf("Date", "Amount", "Currency", "Category", "Note", "Type")) + transactions.map {
            listOf(dateFormatter.format(it.timestamp), it.amount.toPlainString(), it.currencyCode, it.categoryLabel, it.note, if (it.amount.signum() >= 0) "Income" else "Expense")
        }
        return ByteArrayOutputStream().use { bytes -> ZipOutputStream(bytes).use { zip ->
            fun entry(name: String, value: String) { zip.putNextEntry(ZipEntry(name)); zip.write(value.toByteArray()); zip.closeEntry() }
            entry("[Content_Types].xml", """<?xml version="1.0"?><Types xmlns="http://schemas.openxmlformats.org/package/2006/content-types"><Default Extension="rels" ContentType="application/vnd.openxmlformats-package.relationships+xml"/><Default Extension="xml" ContentType="application/xml"/><Override PartName="/xl/workbook.xml" ContentType="application/vnd.openxmlformats-officedocument.spreadsheetml.sheet.main+xml"/><Override PartName="/xl/worksheets/sheet1.xml" ContentType="application/vnd.openxmlformats-officedocument.spreadsheetml.worksheet+xml"/></Types>""")
            entry("_rels/.rels", """<?xml version="1.0"?><Relationships xmlns="http://schemas.openxmlformats.org/package/2006/relationships"><Relationship Id="rId1" Type="http://schemas.openxmlformats.org/officeDocument/2006/relationships/officeDocument" Target="xl/workbook.xml"/></Relationships>""")
            entry("xl/workbook.xml", """<?xml version="1.0"?><workbook xmlns="http://schemas.openxmlformats.org/spreadsheetml/2006/main" xmlns:r="http://schemas.openxmlformats.org/officeDocument/2006/relationships"><sheets><sheet name="Transactions" sheetId="1" r:id="rId1"/></sheets></workbook>""")
            entry("xl/_rels/workbook.xml.rels", """<?xml version="1.0"?><Relationships xmlns="http://schemas.openxmlformats.org/package/2006/relationships"><Relationship Id="rId1" Type="http://schemas.openxmlformats.org/officeDocument/2006/relationships/worksheet" Target="worksheets/sheet1.xml"/></Relationships>""")
            val sheet = rows.mapIndexed { rowIndex, row -> "<row r=\"${rowIndex + 1}\">" + row.mapIndexed { column, value -> "<c r=\"${('A'.code + column).toChar()}${rowIndex + 1}\" t=\"inlineStr\"><is><t>${escape(value)}</t></is></c>" }.joinToString("") + "</row>" }.joinToString("")
            entry("xl/worksheets/sheet1.xml", """<?xml version="1.0"?><worksheet xmlns="http://schemas.openxmlformats.org/spreadsheetml/2006/main"><sheetData>$sheet</sheetData></worksheet>""")
        }; bytes.toByteArray() }
    }
    private fun escape(value: String) = value.replace("&", "&amp;").replace("<", "&lt;").replace(">", "&gt;")
}
